Gems/Atom/Utils/Code/Include/Atom/Utils/StableDynamicArray.h

```diff
@@ -263,6 +270,68 @@ namespace AZ
 
     };
 
+    /**
+    * A weak reference to the data allocated in the array. The ValueType can be forward declared, and is only used for type safety.
+    * Data cannot be accessed directly from a weak handle. To access the data you must go through the array itself.
+    * There is no guarantee that a weak handle is not dangling. In the initial implementation, there is not a way to
+    * check if it is dangling, so it should only be used in situations where it is known that the owning handle
+    * has not gone out of scope or been defragmented. Eventually, this implementation could be augmented with
+    * salt/generationId to verify whether or not the weak handle is dangling.
+    */
+    template<typename ValueType>
+    class StableDynamicArrayWeakHandle
+    {
+    public:
+        StableDynamicArrayWeakHandle() = default;
+        StableDynamicArrayWeakHandle(uint32_t integer)
+        {
+            m_pageIndex = integer >> 16;
+            m_elementIndex = integer & 0xFFFF;
+        }
+
+        uint32_t GetUint() const
+        {
+            uint32_t result = m_pageIndex << 16;
+            result |= static_cast<uint32_t>(m_elementIndex);
+            return result;
+        }
+
+        bool IsValid() const
+        {
+            return m_pageIndex != AZStd::numeric_limits<uint16_t>::max() &&
+                m_elementIndex != AZStd::numeric_limits<uint16_t>::max();
+        }
+
+        bool operator<(const StableDynamicArrayWeakHandle<ValueType>& rhs) const
+        {
+            return GetUint() < rhs.GetUint();
+        }
+        bool operator==(const StableDynamicArrayWeakHandle<ValueType>& rhs) const
+        {
+            return m_pageIndex == rhs.m_pageIndex && m_elementIndex == rhs.m_elementIndex;
+        }
+        bool operator!=(const StableDynamicArrayWeakHandle<ValueType>& rhs) const
+        {
+            return m_pageIndex != rhs.m_pageIndex || m_elementIndex != rhs.m_elementIndex;
+        }
+
+    private:
+        StableDynamicArrayWeakHandle(uint16_t pageIndex, uint16_t elementIndex)
+            : m_pageIndex(pageIndex)
+            , m_elementIndex(elementIndex)
+        {
+        }
+
+        uint16_t m_pageIndex = AZStd::numeric_limits<uint16_t>::max();
+        uint16_t m_elementIndex = AZStd::numeric_limits<uint16_t>::max();
```

Comment:
Seems like a union of a bitfield and a uint32_t would be useful here so you could transparently retrieve the uint32_t value.
```c++
union {
    struct 
    {
        uint32_t m_pageIndex:16;
        uint32_t m_elementIndex:16;
    };
    uint32_t m_value;
};
```